### PR TITLE
Add competitive orders

### LINKED ISSUE #4: Competitive orders
As discussed in TC meeting 1/19, we want to add special *competitive order*. The competitive order is similar to a normal order, but only the team that delivers first will get points for delivering the order.

The rule change proposal should describe:
* how many order may be competitive orders
* what kind of product can be requested (e.g., should we restrict to C0s?)
* how scoring for partial production steps works (i.e., is it all or nothing?)

### PR DESCRIPTION
**Pull request by morxa**
For a competitive order, only the team that delivers first gets points for the delivery, the second team will not get any points.

Points for production steps are unaffected.

This resolves #4.

### CODE REVIEW

**Comment by thomerz** [comment]
I am absolutely no fan of this change proposal since it further complicates our league which makes it even harder to explain, visualize, and so on. Also, this change leads our league into a state where we value fast automation more than collaborative production. For instance, producing two products, C0 (non-competitive) and C1 (competitive) could yield the same points as producing the same C1 first. I do not think that we want this to happen at the moment.

**Comment by morxa** [comment]
I don't see how this values fast automation over collaboration. For more complex products, your team will not get any points if it tries to produce the competitive order with only a single robot, because the other team will be faster.

In your example, the single C1 can already give you more points than the C0 and the C1 with the current rules. If you don't manage to produce the C1 in time, you will not get many points. The only difference is that the deadline is not fixed, but determined by the speed of the other team.

In my opinion, competitive orders would enhance the competition aspect of our league and would make the scenario more realistic.

**Comment by morxa** [comment]
@MostafaGomaa @teamsolidus @vcoelen @chde, any comments?

**Comment by thomerz** [comment]
The same scenario is possible now if you miss the delivery timeslot, that is true. But my scenario is also possible for standing orders. I think we have a lot of other issues in this league than this proposal. As long as no complex products such as C2 or C3 are produced on a regular basis, I personally don't see the need for such a drastic rule change. In our TelCo, we were talking about making 1 C0 order competitive. Your proposal now suggests multiple competitive orders.

In addition, I do not see where a link to real-world production would be? Both teams get a binding contract to deliver certain products. If the product is delivered too late, a penalty is deducted from the agreed price. I cannot imagine any real-world scenario, where the first one delivering a product requested from 2 suppliers would get paid substantially more, than the second one, if both are delivered in the specified time frame.

File: rulebook.tex [hunk 1]
```diff
@@ -1364,32 +1364,41 @@ \subsubsection{Production Plan}
...
-first 300 seconds. An order will be announced before the delivery time
-slots starts. For $C_0$ products the lead time ahead of the start of
-the time window is 60 to 120 seconds, for $C_1$ it is 120 to 300
-seconds, and for $C_2$ and $C_3$ orders the time is between 300 and
-600 seconds. Two early orders will be posted that are announced in the
-first third of the game with a delivery time window in the last
-third. For each of the complexities $C_0$ and $C_1$ a standing order
-for a single product will be placed at the beginning of the production
-phase, which can be fulfilled at any time during the game.
+tournament phase. But the actual production chains are randomized.
```
Comment (timn) [inline rulebook.tex]: What has changed here? Is it just reformatting? In that case, remove. Otherwise, please give a hint.

File: rulebook.tex [hunk 2]
```diff
@@ -1497,6 +1506,12 @@ \subsection{Task Fulfillment and Scoring}
 awarded. Performing a step for a later order which has not yet been
 announced cannot be awarded.
 
+For a competitive order, the same scoring scheme for in-time, delayed,
+late, and wrong delivery applies, with an additional factor of $2$ if
```
Comment (timn) [inline rulebook.tex]: Again, no factor in my opinion.

File: rulebook.tex [hunk 3]
```diff
@@ -1554,6 +1569,10 @@ \subsection{Task Fulfillment and Scoring}
         delivered
         & $+1$\\
         False delivery & Deliver an intermediate product & $0$\\
+        1st competitive delivery & Factor for delivering a product for a competitive
```
Comment (timn) [inline rulebook.tex]: Can be removed if my comment from above is accepted.

**Comment by timn** [comment]
Chiming in late to the discussion, but anyway. We need to be aware that this pushes the scoring scheme even further into a fast-throughput scenario (see [RC2015 paper](https://link.springer.com/chapter/10.1007/978-3-319-29339-4_3)). I do not share @thomerz' fear that this could lead to more automation. Rather, it could lead to more interest in speedy *and robust* production.

But I do think we should approach this carefully. As my review comments suggest, I'd have two competitive orders, one in the main game time and one in overtime. Then, do not score higher, or only mildly higher than a regular product (afterall, you do not have to do more for that product, and the risk is the same for both teams which attempt it, you can also use this as a tactical element to "steal" points if you are fast enough and the other team is also trying).

I do see @morxa real-world scenario. Some car manufacturers operate with a zero-storage factory (actually they have a short buffer, which is why a strike or other influences can lead to lower production so quickly). Now, if you need something urgently, you go to your two or three suppliers and buy where you can get it in time.

So, this might be something to discuss in a TC meeting to exchange arguments and make decisions about the points in question, now that we have a proposal on the table.

File: rulebook.tex [hunk 4]
```diff
@@ -1364,32 +1364,41 @@ \subsubsection{Production Plan}
...
 all. Products retrieved from the Storage Station may be used only for
 $C_0$ orders with a requested quantity equal to or greater than two.
-
-In each game, up to 10 orders will be placed.\footnote{There will up
-  to 9 orders in regular game time, and an additional order in the
-  case of overtime. The number can be adjusted during the tournament
-  by the Technical Committee, should this turn out to be necessary.}
+An order can be \emph{competitive} or \emph{non-competitive}. For a
+competitive order, only the team that delivers first will score points
+for the delivery, and it will score twice the amount of points. The
```
[the remark this answers is not in the document]
Comment (morxa) [reply rulebook.tex]: But on the other hand, if the team scores the same amount of points, the incentive to pursue a competitive order is rather low, because it is more risky and offers no benefit other than blocking the other team from scoring those points. This may be sufficient if there are a number of competitive orders, but if we limit them, then the other team may also just decide to ignore it.

I think pursuing a competitive order should have a benefit other than blocking the other team from scoring.

File: rulebook.tex [hunk 5]
```diff
@@ -1364,32 +1364,41 @@ \subsubsection{Production Plan}
...
+other team will not get any points for the delivery, even if it delivers
+in time.  For a non-competitive order, both teams score on delivery with
+a factor of $1$.
+
+In each game, up to 9 orders will be placed within the regular game.
+4 of those 9 orders will be \emph{competitive}, while the remaining
```
[the remark this answers is not in the document]
Comment (morxa) [reply rulebook.tex]: I'm changing it to 1 + 1 in overtime.

File: rulebook.tex (hunk 1 above)
Comment (morxa) [reply rulebook.tex]: I added the sentence
> A competitive order always requires exactly one product to be delivered.

**Comment by morxa** [comment]
I changed the proposal to 1 competitive order during the regular game + 1 competitive order during overtime.

I applied some of the suggested fixes. As mentioned in the comments, I'd still propose to give more points for a successful competitive order. This increases the incentive to purse such an order and balances out the risk of not scoring at all. Otherwise, I'd expect teams to simply ignore competitive orders.

File: rulebook.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (timn) [reply rulebook.tex]: The incentive is to "steal" the points from the other team. But I do see your point that at the current state some (most?) teams will simply pursue their guaranteed orders.

I think what I really dislike is a "factor" of points. We do not have this anywhere. If we keep this, let's make this a static number of points, e.g., 10 points on successful (accepted) competitive delivery. In the refbox, there is currently no notion of factors, just summed points. I don't think we need to account for late delivery for the competitive points addition, if you deliver and score at all, you get 10 points (or so) extra.

[the remark this answers is not in the document]
Comment (morxa) [reply rulebook.tex]: The idea of the factor was that it makes the rule independent of the actual scoring scheme for the non-competitive variant. But I think we can also rephrase this as only giving additional points for the delivery of a competitive order, and no points for delivery if it is the second delivery for that order.

This would reduce the penalty for a second delivery for a competitive complex order, as the team would still get points for the intermediate steps, just not for the delivery anymore.

[the remark this answers is not in the document]
Comment (morxa) [reply rulebook.tex]: We could also remove the time limit on the competitive order, i.e., make it a standing order. This would also fit with the real-world scenario in mind, where the manufacturer just wants to get the product as fast as possible, but without a specified deadline. What do you think?

This would make the scoring a lot simpler.

[the remark this answers is not in the document]
Comment (morxa) [reply rulebook.tex]: I got rid of the factor and assigned +20 for the first delivery, which doubles the delivery points for a regular delivery.

[the remark this answers is not in the document]
Comment (timn) [reply rulebook.tex]: This is too much in my opinion. It is given too much emphasis on the competitive order. As I stated earlier, let's start slowly and see how it goes. 10 seems to be enough for me, that's ~30% of the overall product you get on top in addition of the fact that a competing team might just have lost work effort worth 30 points if they were slower.

Making it a standalone order makes it less of a strategic run-time decision, I'd keep it with a time window (in overtime the time window would span the entire overtime).

**Comment by MostafaGomaa** [comment]
I am in favor of the proposal on its current form. That is, the competition is restricted to getting some bonus points instead of not scoring at all. That earlier, no scoring at all, was a little too extreme in my opinion. After getting all the production steps right and getting no reward if delivered 2 seconds later. 

- I am definitely in favor of the extra competitive order in over time
- On the points awarded, I think 10 is too little and 20 is about enough to motivate the competition
- I think having it as a standing order does make it, less of a strategic decision and even could be more luck thing  (of course depending on how the teams reason about production goals)

File: rulebook.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (morxa) [reply rulebook.tex]: I reduced the bonus to 10 points.

File: rulebook.tex [hunk 6]
```diff
@@ -1498,32 +1498,41 @@ \subsubsection{Production Plan}
...
 all. Products retrieved from the Storage Station may be used only for
 $C_0$ orders with a requested quantity equal to or greater than two.
-
-In each game, up to 10 orders will be placed.\footnote{There will up
-  to 9 orders in regular game time, and an additional order in the
-  case of overtime. The number can be adjusted during the tournament
-  by the Technical Committee, should this turn out to be necessary.}
+An order can be \emph{competitive} or \emph{non-competitive}. For a
+competitive order, only the team that delivers first will score points
+for the delivery, and it will score twice the amount of points. The
```
Comment (timn) [inline rulebook.tex]:
```suggestion
for the delivery, and it will score higher (cf.~\reftab{tab:scoring-production}). The
```
This is wrong. Delivery in time 20 + 1st competitive 10 = 30, which is not twice the amount of points. This should just state "scores higher" and leave the actual value to the table, as to avoid inconsistencies if there are multiple sources of information for the same thing (we had plenty of those in the early years).

**Comment by thomerz** [comment]
+1

File: rulebook.tex (hunk 6 above)
Comment (morxa) [reply rulebook.tex]: Thanks for spotting this :+1:

**Comment by timn** [comment]
There is one more thing that I noticed when reading it again: it might be worth specifically pointing out that points for production steps will be given. That's how you have to read the current text, since it specifically says "will not get any points for the delivery". If the intention is not even to give production points clarification is needed even more so. But I think in anticipation of competitive orders which may involve, e.g., a C1, no delivery points would already be penalty enough. What do you think?

**Comment by morxa** [comment]
> There is one more thing that I noticed when reading it again: it might be worth specifically pointing out that points for production steps will be given. That's how you have to read the current text, since it specifically says "will not get any points for the delivery". If the intention is not even to give production points clarification is needed even more so. But I think in anticipation of competitive orders which may involve, e.g., a C1, no delivery points would already be penalty enough. What do you think?

Yes, that's the intention. Originally, the idea was also to give production points only for the first delivery, but we decided against that in a TC meeting. Now, only the points for delivery depend on if you're first or not.

**Comment by timn** [comment]
Ok, that must have been the meeting I could not attend. Then we should clarify this by stating explicitly that production step points will still be awarded.

**Comment by morxa** [comment]
> Ok, that must have been the meeting I could not attend. Then we should clarify this by stating explicitly that production step points will still be awarded.

Done :+1: Please check again.

File: rulebook.tex [hunk 7]
```diff
@@ -1630,6 +1640,10 @@ \subsection{Task Fulfillment and Scoring}
 awarded. Performing a step for a later order which has not yet been
 announced cannot be awarded.
 
+For a competitive order, the same scoring scheme for in-time, delayed, and late
+delivery applies for the first delivery, with an addition of $10$ points for a
```
Comment (timn) [inline rulebook.tex]: This explicit score should be removed here, point to the table, reduce redundancies especially for the crucial scoring bit (we ran into ambiguities in the past if only one place was updated and the other forgotten).

Comment (morxa) [reply rulebook.tex]: Makes sense. I did not change the sentence about giving no points to the second delivery, as this will most probably not change, and it helps to understand competitive orders.

**Comment by MostafaGomaa** [comment]
Sorry, I am a bit confused of how the thread developed!
In my opinion, the difference of 30 points between the first and second delivery is way too much.
I was on board with the proposal at a point I thought we agreed that both production and delivery points should be awarded normally for both 1st and 2nd delivery and that the competition element is defined by an extra  +20 points, awarded for the team that scores first.
The benefits of that as I saw it :
 1 - It is easier to formulate and understand
          (ie, *  Scoring Table: Needs to mentions only one row about awarding of the competition points,
                   as opposed to, having 2 'extra' different types of delivery, one of them pointing to originally
                   existing delivery and the other one to assert a zero score.
                * Text: Nothing needs to be mentioned about 'not scoring')
 2 - It does not deprive the 2nd team of well-deserved points of delivery, in case they attempt to deliver.
       (My reasoning here, Is that delivery points are awarded for completing the last step of production by moving the caped workpiece from the CS to the DS and delivering on time and partial scoring for late delivery. A team that scores second _in my opinion_ still deserves the points giving for this manipulation.  It would be  _in my opinion_ a bit extream to get nothing at all after successfully performing those steps)
 3 -  It clearly separates the delivery concept of the competitiveness concept
       (This is a bit redundant with point 2, yet I mean it here in a more general sense. For instance, We can more smoothly, later on, extend the competitiveness concept to other aspects of the game with minimal changes to the existing concepts. Possible use-cases, making exploration competitive, making scoring on production steps competitive pushing further the concept of a competitive order for higher complexity products)

That said, even though I prefer the 'Bonus' formulation, I am not completely against formulating the difference in scoring between 1st and 2nd scorer as 'Deducing Delivery Points' for the later. After all, it is another way to assert the difference in scoring. 
In that case, my recommendation would be:
- Communicate 1st scoring to all soon
  (We would need to make sure to publish the information on the public channel, as soon as a 1st delivery had occurred!. That is, to refrain from forcing the late scorers from wasting time on an eventually useless delivery, in case they wish to abandon it. If we agree on that point, we should probably mention that in the rules.)
-  No extra awarded points only deduction
 (A difference of 30 points is too extreme in my opinion. The extra +10 award could be abandoned. The difference of +20 for the 1st scorer, would already be ensured by deducing the delivery)

Sorry for the lengthy text!

**Comment by timn** [comment]
While I'm fine with the existing formulation, I do support @MostafaGomaa's statement and would prefer the "bonus-only" version.

**Comment by timn** [comment]
> I appreciate feedback, but it's rather late for a lengthy discussion of the proposal. It's been up here since October and we want to finish the rulebook soon.

That should not prevent having a discussion. It is only an indicator to handle this a little more quickly.

My (abbreviated) understanding would be this:
- do not score zero on second delivery, i.e., we would have to write nothing about this at all in the rules
- get bonus for first delivery, e.g., 20 points

The latter allows to condense the text tremendously and avoids the few excerpts that can make this harder to understand than necessary, i.e., nothing changes for the production and delivery scoring. The only thing you need to explain is that the first to deliver gets a bonus. It also avoids a "penalty". However, if you ground this in an industrial logistics process you would get zero when delivering later, as you loose the business to your competitor.

As I mentioned before, I think both ways are fine, the bonus-only way may be simpler to convey and a safer version to start from. My two cents.

**Comment by dka14** [comment]
To shorten the discussion I suggest the following wording:
In case of a competive order the team with first (correct) delivery gets a bonus of 20 points.

**Comment by thomerz** [comment]
I agree with @timn and @dka14 to have (at least for this year) the simpler of the two options where the first delivery scores an additional +20 points, whereas the second delivery scores standard points. We agreed at the first TC meeting to basically not touch the scoring scheme. The second option, in my opinion, adds too much complexity, e.g. we have to notify teams of a successful first delivery such that the second team can cancel production of that order early.

Also since this rule change is already labelled for voting, I would suggest to remove that and post a summary of the proposed rule change such that a summary of this discussion can be seen on which the vote is based.

**Comment by morxa** [comment]
The bonus-only version remove a good part of the competition factor of the proposal. With the current proposal, deciding whether to pursue a competitive order or an alternative non-competitive order (with the same complexity) actually involves some reasoning, for a very simple reason:
* If you win the competitive order, you get more points
* If you lose the competitive order, you get less points

This leads to interesting questions about the decision-making process and introduces some game-theoretic reasoning into the league: You may want to estimate the performance of the other team in order to make a good decision. You can also try to guess whether the other team is trying to accomplish the competitive order or not, as this influences your decision.

These game-theoretic strategy questions do not come up with the two proposed alternatives:
* If we remove the bonus, then there is only a much smaller advantage (as explained above). In that case, competitive orders become very unattractive.
* If we only give bonus points but also give points for the second delivery, there is no reason not to pursue a competitive order. The decision-making process is trivial. You can completely ignore whether an order is competitive or not, and still end up with an optimal strategy in most situations. You can also completely ignore the other team.

At the same time, with the current proposal, you *don't have to* do any game-theoretic reasoning. It may give a strategic advantage, but it's not essential to the game at all.

> The latter allows to condense the text tremendously

This is not true. First of all, we would still need to explain what a competitive order is, and that there is a distinction between first and second delivery. Second, the proposal only adds ~6 sentences and two rows to the scoring table to begin with. That's not a tremendous amount of text. In fact, I think the proposed change introduces new, interesting concepts to the competition (see above) without a lot of new text.

> It also avoids a "penalty".

There is no penalty, you don't get any points deducted.

> We agreed at the first TC meeting to basically not touch the scoring scheme and complexity of the game. 

I don't remember that we agreed not to change the scoring scheme, and the notes of the first meeting also contain a section about the competitive orders, which says: *Only first team making the delivery gets the points*. Nobody objected at that point. But the scoring scheme also doesn't change drastically, it only involves a distinction between first and second delivery (which is necessary even with the bonus-only version). The change in complexity is rather benign. The main idea of competitive orders is rather easy to understand, the game-theoretic implications (see above) can always be ignored. You can come up with a very simple strategy that works, but you could also invest considerable efforts in analyzing the opponent to come up with a better strategy if you're interested (I wouldn't expect any team to do this due to the low benefit of 10 points, but that may change in the future). That being said, the change of *complexity of the rules* is rather small. 

> The second option, in my opinion, adds too much complexity, e.g. we have to notify teams of a successful first delivery such that the second team can cancel production of that order early.

This is still necessary for the bonus-only version and very simple to implement. We already share the order info including the delivered orders anyway.

In my opinion, the bonus-only version is rather pointless, as it removes much of the competition factor. The only thing that remains is that a team needs to be faster than the other team to get the bonus points. This may involve some re-prioritization within a team's tasks, but I'd expect the overall effect to be minimal. In that case, I'd rather not have competitive orders at all, because it adds complexity to the rules without adding much to the game.

**Comment by timn** [comment]
So there are now three proposals:
- no delivery points for second and bonus points for first
- delivery points as usual, additional bonus for first
- no competitive orders at all

I think this is a case where we should discuss this (virtual) face to face in a meeting. That seems to be the more productive way to resolve this. @morxa, can you schedule a meeting?

**Comment by morxa** [comment]
I added the discussed changes. I also added that the point deduction for the second delivery may not exceed the points given for the delivery (e.g., in case of a late delivery).

Please review again, thank you!

**Comment by morxa** [comment]
Can someone please review the updated proposal, so we can start voting on it? Thanks!

File: rulebook.tex [hunk 8]
```diff
@@ -1630,6 +1639,15 @@ \subsection{Task Fulfillment and Scoring}
 awarded. Performing a step for a later order which has not yet been
 announced cannot be awarded.
 
+For a competitive order, the same scoring scheme for in-time, delayed, and late
```
Comment (timn) [inline rulebook.tex]: What's the difference between delayed and late delivery?

File: rulebook.tex [hunk 9]
```diff
@@ -1630,6 +1639,15 @@ \subsection{Task Fulfillment and Scoring}
...
+delivery applies, with additional points for a successful first delivery
+and a point deduction for the second delivery (cf.~\reftab{tab:scoring}). The
+deduction may not exceed the points given for the delivery.\footnote{As an
```
Comment (timn) [inline rulebook.tex]: Place the "." after the footnote. We use the fnpct package which makes the display nicer, but it requires points to be set after the footnote (in LaTeX you seemingly can only do lookahead, but not look back).
```suggestion
deduction may not exceed the points given for the delivery\footnote{As an
```

File: rulebook.tex [hunk 10]
```diff
@@ -1630,6 +1639,15 @@ \subsection{Task Fulfillment and Scoring}
...
+  example, if a team delivers late and scores only $5$ points for the delivery,
+only $5$ points will be deducted.} Points for production steps are given in the
```
Comment (timn) [inline rulebook.tex]:
```suggestion
only $5$ points will be deducted.}. Points for production steps are given in the
```

File: rulebook.tex [hunk 11]
```diff
@@ -1729,6 +1747,16 @@ \subsection{Task Fulfillment and Scoring}
...
         False delivery & Deliver an intermediate product & $0$
         \\
+        1st competitive delivery
+        & Points for the first delivery for a competitive order. The score is
+        given in addition to the points for the regular delivery.
+        &$+ 10$
+        \\
+        2nd competitive delivery
+        & Point deduction for the second delivery for a competitive order. The
+        points are deducted from the delivery points.
```
Comment (timn) [inline rulebook.tex]: Here would be a good place to state once more that zero is the minimum. This is what teams may look at for scores more frequently.

File: rulebook.tex (hunk 8 above)
Comment (morxa) [reply rulebook.tex]: Delayed: within 10 seconds after the deadline, late: more than 10 seconds after the deadline (as specified in the scoring table)

File: rulebook.tex (hunk 11 above)
Comment (morxa) [reply rulebook.tex]: Yes, makes sense, added.

File: rulebook.tex (hunk 10 above)
Comment (morxa) [reply rulebook.tex]: Fixed, thanks for the suggestion

**Comment by morxa** [comment]
@timn thanks for the suggestions and questions, all of them should be answered now.

File: rulebook.tex (hunk 8 above)
Comment (timn) [reply rulebook.tex]: Ah, got ya.

**Comment by morxa** [comment]
Voting opened. Please add your vote!

**Comment by timn** [comment]
+1

**Comment by thomerz** [comment]
+1

**Comment by teamsolidus** [comment]
+1

**Comment by morxa** [comment]
+1

**Comment by morxa** [comment]
Please vote today if you still want to vote!

**Comment by morxa** [comment]
**Proposal accepted**

* in favor: 4
* abstentions: 3
* opposed: 0
